geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSet.java

```diff
@@ -300,7 +283,7 @@ private void srandomDuplicateList(int count, List<byte[]> result) {
     }
   }
 
-  private void srandomUniqueListWithSmallCount(int count, List<byte[]> result) {
+  private void srandomUniqueListAddToResult(int count, List<byte[]> result, boolean isSPop) {
```

Comment:
I think it would be better to leave this method and `srandomUniqueListWithLargeCount()` as they were and introduce new logic specifically for SPOP. Although the process for SPOP and SRANDMEMBER is broadly the same, there's a key difference, which is that SPOP modifies the set as it goes, which could lead to rehashing of the backing array, which in turn leads to `indexesUsed` becoming incorrect. Since with SPOP we can't possibly remove the same element twice, it's not necessary to keep track of the indexes used, and we can instead just check if the member at the random index we picked was null or not:
```
  public List<byte[]> spop(int count, Region<RedisKey, RedisData> region, RedisKey key) {
    List<byte[]> result = new ArrayList<>();
    if (members.size() <= count) {
      result.addAll(members);
      members.clear();
    } else {
      Random rand = new Random();
      while (result.size() != count) {
        int randIndex = rand.nextInt(members.getBackingArrayLength());
        byte[] member = members.getFromBackingArray(randIndex);
        if (member != null) {
          result.add(member);
          members.remove(member);
        }
      }
    }

    storeChanges(region, key, new RemoveByteArrays(result));
    return result;
  }
```

Reply:
It only adds the index if it isn't spop. When it does spop there is an indexUsed set, but nothing gets added so it is a waste.

Reply:
Ah, true. The behaviour would be correct with SPOP, but as you say, we're creating a HashSet that we don't need to here.